### PR TITLE
add AdvancedStatefulSet feature

### LINKED ISSUE #1094: Use AdvancedStatefulSet CRD to replace Kubernetes StatefulSet
## Feature Request

**Is your feature request related to a problem? Please describe:**

https://github.com/pingcap/tidb-operator/issues/981

**Describe the feature you'd like:**

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

https://github.com/cofyc/advanced-statefulset

### PR DESCRIPTION
**Pull request by cofyc**
### What problem does this PR solve? 

fixes #1094

### Check List 

Tests 

 - Unit test
 - E2E test
 - Stability test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has Helm charts change
 - Has Go code change
 - Has CI related scripts change
 - Has documents change

Side effects

```
tidb-operator chart value `scheduler.features` is migrated to `features`
```

### Does this PR introduce a user-facing change?:
 
 ```release-note
add AdvancedStatefulset feature to use advanced statefulset instead of Kubernetes builtin statefulset
 ```

### CODE REVIEW

File: charts/tidb-operator/templates/advanced-statefulset-deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    matchLabels:
+      app: advanced-statefulset-controller
+  template:
+    metadata:
+      labels:
+        app: advanced-statefulset-controller
+    spec:
+      containers:
+      - name: advanced-statefulset-controller
+        image: quay.io/cofyc/advanced-statefulset:latest
```
Comment (gregwebs) [inline charts/tidb-operator/templates/advanced-statefulset-deployment.yaml]: Can we put this under pingcap now? Including the github repo?

Comment (cofyc) [reply charts/tidb-operator/templates/advanced-statefulset-deployment.yaml]: yes, when this solution works well

File: charts/tidb-operator/templates/advanced-statefulset-deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,24 @@
+apiVersion: apps/v1
```
Comment (cofyc) [inline charts/tidb-operator/templates/advanced-statefulset-deployment.yaml]: I'll add an option in helm chart to deploy advanced statefulset when the feature is enabled.

File: charts/tidb-operator/templates/advanced-statefulset-rbac.yaml [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  resources:
+  - 'endpoints'
+  verbs:
+  - '*'
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: advanced-statefulset-controller
+  namespace: kube-system
```
Comment (tennix) [inline charts/tidb-operator/templates/advanced-statefulset-rbac.yaml]: Why this ns is `kube-system`?

Comment (cofyc) [reply charts/tidb-operator/templates/advanced-statefulset-rbac.yaml]: testing only, this controller does not support specifying namespace (working on it)

File: hack/kind-cluster-build.sh [hunk 4]
```diff
@@ -62,7 +62,7 @@ done
 clusterName=${clusterName:-kind}
 nodeNum=${nodeNum:-6}
 k8sVersion=${k8sVersion:-v1.12.8}
-volumeNum=${volumeNum:-9}
+volumeNum=${volumeNum:-32}
```
Comment (cofyc) [inline hack/kind-cluster-build.sh]: able to run e2e tests in a small cluster

File: charts/tidb-operator/templates/advanced-statefulset-deployment.yaml (hunk 2 above)
Comment (cofyc) [reply charts/tidb-operator/templates/advanced-statefulset-deployment.yaml]: done

File: charts/tidb-operator/values.yaml [hunk 5]
```diff
@@ -63,8 +80,6 @@ scheduler:
   logLevel: 2
   replicas: 1
   schedulerName: tidb-scheduler
-  # features:
-  # - StableScheduling=true
```
Comment (cofyc) [inline charts/tidb-operator/values.yaml]: this feature is enabled by default and I guess no one will disable it in production, so I think we don't need to support `scheduler.features` for backward compatibility

File: charts/tidb-operator/values.yaml [hunk 6]
```diff
@@ -18,6 +18,23 @@ defaultStorageClassName: local-storage
...
+#   AdvancedStatefulSet (default: false)
+#     If enabled, tidb-operator will use AdvancedStatefulSet to manage pods
+#     instead of Kubernetes StatefulSet.
+#     It's ok to turn it on if this feature is not enabled. However it's not ok
+#     to turn it off when the tidb-operator already uses AdvancedStatefulSet to
+#     manage pods. This is in alpha phrase.
+#
+features: {}
+# - AdvancedStatefulSet=false
+# - StableScheduling=true
```
Comment (cofyc) [inline charts/tidb-operator/values.yaml]: some features involve multiple components, it's better to make it global

File: tests/manifests/e2e/e2e.yaml [hunk 7]
```diff
@@ -47,8 +47,8 @@ spec:
     command:
     - /usr/local/bin/e2e
     - --operator-tag=e2e
-    - --operator-image=pingcap/tidb-operator:latest
-    - --test-apiserver-image=pingcap/test-apiserver:latest
+    - --operator-image=localhost:5000/pingcap/tidb-operator:latest
+    - --test-apiserver-image=localhost:5000/pingcap/tidb-operator:latest
```
Comment (cofyc) [inline tests/manifests/e2e/e2e.yaml]: e2e file only, I think it's fine to use local images by default

File: charts/tidb-operator/templates/advanced-statefulset-rbac.yaml (hunk 3 above)
Comment (cofyc) [reply charts/tidb-operator/templates/advanced-statefulset-rbac.yaml]: updated

File: cmd/admission-controller/main.go [hunk 8]
```diff
@@ -68,10 +72,22 @@ func main() {
...
 	}
+	asCli, err := asclientset.NewForConfig(cfg)
+	if err != nil {
+		glog.Fatalf("failed to get advanced-statefulset Clientset: %v", err)
+	}
+
+	if features.DefaultFeatureGate.Enabled(features.AdvancedStatefulSet) {
+		// If AdvancedStatefulSet is enabled, we hijack the Kubernetes client to use
+		// AdvancedStatefulSet.
+		kubeCli = helper.NewHijackClient(kubeCli, asCli)
```
Comment (cofyc) [inline cmd/admission-controller/main.go]: https://github.com/cofyc/advanced-statefulset/blob/master/pkg/apis/apps/v1alpha1/helper/hijack.go

File: tests/manifests/e2e/e2e.yaml (hunk 7 above)
Comment (aylei) [inline tests/manifests/e2e/e2e.yaml]:
```suggestion
    - --test-apiserver-image=localhost:5000/pingcap/test-apiserver:latest
```

File: charts/tidb-operator/values.yaml (hunk 5 above)
Comment (aylei) [reply charts/tidb-operator/values.yaml]: Breaking backward compatibility should be noted in the PR description.

File: charts/tidb-operator/templates/advanced-statefulset-deployment.yaml [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        app.kubernetes.io/instance: {{ .Release.Name }}
+        app.kubernetes.io/component: advanced-statefulset-controller
+    spec:
+      serviceAccountName: advanced-statefulset-controller
+      containers:
+      - name: advanced-statefulset-controller
+        image: {{ .Values.advancedStatefulset.image }}
+        imagePullPolicy: {{ .Values.advancedStatefulset.imagePullPolicy | default "IfNotPresent" }}
+        args:
+        - --v=4
```
Comment (onlymellb) [inline charts/tidb-operator/templates/advanced-statefulset-deployment.yaml]: The log level of all components we deployed is set to 2 by default. Is it set to 2 by default?

File: charts/tidb-operator/templates/controller-manager-rbac.yaml [hunk 10]
```diff
@@ -56,6 +56,12 @@ rules:
   - restores
   - restores/finalizers
   verbs: ["*"]
+- apiGroups:
+  - apps.pingcap.com
+  resources:
+  - statefulsets
+  verbs:
+  - '*'
```
Comment (onlymellb) [inline charts/tidb-operator/templates/controller-manager-rbac.yaml]: Should this be configured only when the advance sts feature is turned on, keeping the principle of least privilege?

File: charts/tidb-operator/templates/controller-manager-rbac.yaml [hunk 11]
```diff
@@ -129,6 +135,12 @@ rules:
...
```
Comment (onlymellb) [inline charts/tidb-operator/templates/controller-manager-rbac.yaml]: ditto

File: charts/tidb-operator/templates/controller-manager-rbac.yaml (hunk 10 above)
Comment (cofyc) [reply charts/tidb-operator/templates/controller-manager-rbac.yaml]: done

File: charts/tidb-operator/templates/controller-manager-rbac.yaml (hunk 11 above)
Comment (cofyc) [reply charts/tidb-operator/templates/controller-manager-rbac.yaml]: done

File: charts/tidb-operator/templates/advanced-statefulset-deployment.yaml (hunk 9 above)
Comment (cofyc) [reply charts/tidb-operator/templates/advanced-statefulset-deployment.yaml]: fixed, add a logLevel field to specify log level (defaults to 2) same as tidb-controller-manager

File: charts/tidb-operator/values.yaml [hunk 12]
```diff
@@ -124,3 +139,36 @@ apiserver:
...
   #   effect: "NoSchedule"
 
+# When AdvancedStatefulSet feature is enabled, you must install
+# AdvancedStatefulSet controller.
+# Note that AdvancedStatefulSet CRD must be installed manually via the following
+# command:
+#   kubectl apply -f manifests/advanced-statefulset-crd.v1beta1.yaml
+advancedStatefulset:
+  create: false
+  image: quay.io/cofyc/advanced-statefulset:latest
```
[the remark this answers is not in the document]
Comment (cofyc) [reply charts/tidb-operator/values.yaml]: I'm going to migrate cofyc/advanced-statefulset repo to pingcap/advanced-statefulset

**Comment by cofyc** [comment]
I'll push an image under pingcap and update this PR.

File: charts/tidb-operator/values.yaml [hunk 13]
```diff
@@ -127,3 +142,38 @@ apiserver:
...
+  replicas: 1
+  resources:
+    limits:
+      cpu: 500m
+      memory: 300Mi
+    requests:
+      cpu: 200m
+      memory: 50Mi
+  # This will default to matching your kubernetes version
+  # kubeSchedulerImageTag:
```
Comment (DanielZhangQD) [inline charts/tidb-operator/values.yaml]: Should these comments be removed?

Comment (cofyc) [reply charts/tidb-operator/values.yaml]: yes, thanks

Comment (cofyc) [reply charts/tidb-operator/values.yaml]: fixed

**Comment by cofyc** [comment]
@tennix @weekface @DanielZhangQD @onlymellb PTAL again.

**Pull request merged**